internal/customgitignore/dir_test.go

```diff
@@ -0,0 +1,732 @@
+	}
+}
+
+func TestRecursivelyParsingGitignoreFilesFromPlainDir(t *testing.T) {
+	t.Parallel()
+
+	plainDir := setupPlainDirWithGitignores(t)
+
+	// Read this dir-tree using customgitignore
+	patterns, _, err := customgitignore.ParseGitIgnores(plainDir, true)
+	if err != nil {
+		t.Fatalf("could not read gitignore patterns for test: %v", err)
+	}
+
+	// expect gitignore.Pattern[] to be empty, meaning no .gitignores were processed
+	if len(patterns) != 0 {
+		t.Errorf("Expected patterns slice read from to be empty, non git-repo," +
+			"because .gitignores are meaningless")
+	}
+}
+
+func TestNonRecursivelyParsingGitignoreFilesFromPlainDir(t *testing.T) {
+	t.Parallel()
+
+	plainDir := setupPlainDirWithGitignores(t)
+
+	// Read this dir-tree using customgitignore
+	patterns, _, err := customgitignore.ParseGitIgnores(plainDir, false)
+	if err != nil {
+		t.Fatalf("could not read gitignore patterns for test: %v", err)
+	}
+
+	// expect gitignore.Pattern[] to be empty, meaning no .gitignores were processed
+	if len(patterns) != 0 {
+		t.Errorf("Expected patterns slice read from to be empty, non git-repo," +
+			"because .gitignores are meaningless")
+	}
+}
+
+func TestParsingGitRepoWithoutGitignoreFiles(t *testing.T) {
+	t.Parallel()
+
+	gitRepo := setupGitRepo(t)
+
+	// context: the repo doesn't have a repo-wide gitignore file
+	repoExcludeFile := filepath.Join(gitRepo, filepath.FromSlash(".git/info/exclude"))
+	os.Remove(repoExcludeFile)
+
+	// context: the dir has been crawled, and all its .gitignores removed
+	err := filepath.WalkDir(gitRepo, func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+
+		if !d.IsDir() && d.Name() == ".gitignore" {
+			os.Remove(path)
+		}
+
+		return nil
+	})
+	if err != nil {
+		t.Fatalf("Could not walk tree removing .gitignores because: %v", err)
+	}
+
+	// context: reading this dir-tree using customgitignore, starting midway
+	// up the tree at ./dir_a
+	start := filepath.Join(gitRepo, "dir_a")
+	patterns, _, err := customgitignore.ParseGitIgnores(start, true)
+
+	// expect the parser to handle these missing files
+	if err != nil {
+		t.Fatalf("customgitignore.ParseGitIgnores should have worked but instead it failed with: %v", err)
+	}
+
+	// expect gitignore.Pattern[] to be empty, meaning no .gitignores were processed
+	if len(patterns) != 0 {
+		t.Errorf("Expected patterns slice read from to be empty")
+	}
+}
+
+func setupGitRepo(t *testing.T) string {
+	t.Helper()
+
+	gitRepo := setupPlainDirWithGitignores(t)
+
+	// initialise a git repo
+	if _, err := git.PlainInit(gitRepo, false); err != nil {
+		t.Fatalf("could not initialise git repo for test: %v", err)
+	}
+
+	return gitRepo
+}
+
+func setupPlainDirWithGitignores(t *testing.T) string {
+	t.Helper()
+
+	// A unique tempdir for local funcs to create a git repo inside
+	// TODO: convert `t.TempDir()` to `testutility.CreateTestDir(t)` once PR #832 lands
+	dir := filepath.Join(t.TempDir(), "base_dir")
```

Comment:
#832 has been merged so this can be actioned now, though I don't think it should block landing this